### PR TITLE
PeekStageFinish without Optimizer

### PR DESCRIPTION
**Pull request by antiguru**
No consumers of PeekStageFinish require the optimizer, a heavy-weight object that holds on the catalog. This allows us to drop the catalog earlier and outside the coordinator main loop.

Part of #28720.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/adapter/src/optimize/peek.rs [hunk 1]
```diff
@@ -49,7 +49,7 @@ pub struct Optimizer {
     /// A snapshot of the cluster that will run the dataflows.
     compute_instance: ComputeInstanceSnapshot,
     /// Optional row-set finishing to be applied to the final result.
-    finishing: RowSetFinishing,
+    pub finishing: RowSetFinishing,
```
Comment (teskje) [inline src/adapter/src/optimize/peek.rs:52]: It's nice to have this private so nobody can muck with the optimizer's internal state and potentially break its assumptions. Can we leave this non-pub and use `optimizer.finishing().clone()` instead? Or, if we are worried about the cost of cloning, have an fn into_finishing(self) -> RowSetFinishing`?

**Review by teskje — APPROVED** [review]
LGTM, except that it feels bad to expose some of the optimizers internal state while it still might be used by the optimizer.

**Review by ggevay — APPROVED** [review]
LGTM (but same comment as @teskje)

File: src/adapter/src/optimize/peek.rs (hunk 1 above)
Comment (antiguru) [reply src/adapter/src/optimize/peek.rs]: Makes sense, backed out that change and reverted to `optimizer.finishing().clone()`.

**Pull request merged**